### PR TITLE
layout: Rewrite whitespace stripping.

### PR DESCRIPTION
**Pull request by pcwalton**
This patch makes Servo unconditionally strip whitespace before text run
scanning (assuming that the `white-space` property allows it). Whitespace
stripping during reflow is now only used for handling whitespace at the ends of
lines; reflow now never attempts to handle ignorable whitespace.

Many CSS tests pass now. There are some new failures, however.

The following reference tests now fail due to a pre-existing bug whereby
whitespace is used to calculate the position of inline hypothetical boxes for
elements with `display: inline; position: absolute`:
- `absolute-replaced-height-036.htm`
- `vertical-align-sub-001.htm`
- `vertical-align-super-001.htm`

The following reference tests fail due to a pre-existing bug whereby we don't
handle `font-size: 0` properly in inline reflow:
- `font-size-zero-1.htm`
- `font-size-zero-2.htm`

The following reference test fails due to the fact that it relied on our
incorrect insertion of whitespace to make room for the black background:
- `inline-formatting-context-007.htm`

r? @mbrubeck

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7024)

### CODE REVIEW

**Comment by pcwalton** [comment]
This patch improves GitHub, Hacker News, and probably most other sites.

**Comment by Ms2ger** [comment]
The changes to `tests/wpt/css-tests/css21_dev/html4/block-in-inline-008.htm` were probably unintentional.

**Comment by jdm** [comment]
Why is inline-formatting-context-007.htm only marked failing on linux x64 in release builds?

**Comment by pcwalton** [comment]
@jdm I think `mach update-css` notices pre-existing platform-specific failures and assumes failures from the logs are platform-specific. Fixed.

File: components/layout/fragment.rs [hunk 1]
```diff
@@ -2102,35 +2106,144 @@ impl Fragment {
...
+                    CharIndex(leading_whitespace_character_count),
+                    -CharIndex(leading_whitespace_character_count));
+
+                return WhitespaceStrippingResult::RetainFragment
+            }
+            SpecificFragmentInfo::UnscannedText(ref mut unscanned_text_fragment_info) => {
+                let mut new_text_string = String::new();
+                let mut modified = false;
+                for (i, character) in unscanned_text_fragment_info.text.char_indices() {
+                    if gfx::text::util::is_bidi_control(character) {
```
Comment (mbrubeck) [inline components/layout/fragment.rs]: Should this use `is_always_discardable_char` (which in theory will also discard soft hyphens)?

**Comment by jdm** [comment]
```
/css21_dev/html4/c5502-imrgn-r-003.htm
--------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/insert-inline-in-blocks-n-inlines-begin-003.htm
----------------------------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/insert-inline-in-blocks-n-inlines-middle-003.htm
-----------------------------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/margin-collapse-137.htm
----------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/margin-collapse-138.htm
----------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-collapsing-bidi-003.htm
----------------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-normal-003.htm
-------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-normal-004.htm
-------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-pre-001.htm
----------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-pre-002.htm
----------------------------------------
PASS expected FAIL [Parent]
```

**Comment by mbrubeck** [comment]
Slightly different results on the Mac builder:

```
/css21_dev/html4/insert-inline-in-blocks-n-inlines-begin-003.htm
----------------------------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/insert-inline-in-blocks-n-inlines-middle-003.htm
-----------------------------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-collapsing-bidi-003.htm
----------------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-normal-003.htm
-------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-normal-004.htm
-------------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-pre-001.htm
----------------------------------------
PASS expected FAIL [Parent]
/css21_dev/html4/white-space-pre-002.htm
----------------------------------------
PASS expected FAIL [Parent]
```

**Pull request merged**